nitransforms/nonlinear.py

```diff
@@ -90,70 +118,162 @@ def from_filename(cls, filename, fmt="X5"):
-#                 'not match the number of dimensions')
-
-#         self._coeffs = np.asanyarray(coefficients.dataobj)
-#         self._knots = ImageGrid(four_to_three(coefficients)[0])
-#         self._cache_moving()
-
-#     def _cache_moving(self):
-#         self._moving = np.zeros((self.reference.shape) + (3, ),
-#                                 dtype='float32')
-#         ijk = np.moveaxis(self.reference.ndindex, 0, -1).reshape(-1, self.ndim)
-#         xyz = np.moveaxis(self.reference.ndcoords, 0, -1).reshape(-1, self.ndim)
-#         print(np.shape(xyz))
-
-#         for i in range(np.shape(xyz)[0]):
-#             print(i, xyz[i, :])
-#             self._moving[tuple(ijk[i]) + self.__s] = self._interp_transform(xyz[i, :])
-
-#     def _interp_transform(self, coords):
-#         # Calculate position in the grid of control points
-#         knots_ijk = self._knots.inverse.dot(np.hstack((coords, 1)))[:3]
-#         neighbors = []
-#         offset = 0.0 if self._order & 1 else 0.5
-#         # Calculate neighbors along each dimension
-#         for dim in range(self.ndim):
-#             first = int(np.floor(knots_ijk[dim] + offset) - self._order // 2)
-#             neighbors.append(list(range(first, first + self._order + 1)))
-
-#         # Get indexes of the neighborings clique
-#         ndindex = np.moveaxis(
-#             np.array(np.meshgrid(*neighbors, indexing='ij')), 0, -1).reshape(
-#             -1, self.ndim)
-
-#         # Calculate the tensor B-spline weights of each neighbor
-#         # weights = np.prod(vbspl(ndindex - knots_ijk), axis=-1)
-#         ndindex = [tuple(v) for v in ndindex]
-
-#         # Retrieve coefficients and deal with boundary conditions
-#         zero = np.zeros(self.ndim)
-#         shape = np.array(self._knots.shape)
-#         coeffs = []
-#         for ijk in ndindex:
-#             offbounds = (zero > ijk) | (shape <= ijk)
-#             coeffs.append(
-#                 self._coeffs[ijk] if not np.any(offbounds)
-#                 else [0.0] * self.ndim)
-
-#         # coords[:3] += weights.dot(np.array(coeffs, dtype=float))
-#         return self.reference.inverse.dot(np.hstack((coords, 1)))[:3]
-
-#     def _map_voxel(self, index, moving=None):
-#         """Apply ijk' = f_ijk((i, j, k)), equivalent to the above with indexes."""
-#         return tuple(self._moving[index + self.__s])
+
+class BSplineFieldTransform(TransformBase):
+    """Represent a nonlinear transform parameterized by BSpline basis."""
+
+    __slots__ = ['_coeffs', '_knots', '_weights', '_order', '_moving']
+
+    def __init__(self, coefficients, reference=None, order=3):
+        """Create a smooth deformation field using B-Spline basis."""
+        super(BSplineFieldTransform, self).__init__()
+        self._order = order
+
+        coefficients = _ensure_image(coefficients)
+
+        self._coeffs = np.asanyarray(coefficients.dataobj)
+        self._knots = ImageGrid(four_to_three(coefficients)[0])
```

Comment:
Seems wasteful to create a bunch of new images to take the first and throw it away almost immediately. ImageGrid already resets to `shape = shape[:3]` if `ndim == 4`.

```suggestion
        self._knots = ImageGrid(coefficients)
```